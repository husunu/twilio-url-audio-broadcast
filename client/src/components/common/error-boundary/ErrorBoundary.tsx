import React, { Component } from 'react';
import { Alert, AlertType } from '@/components/common/alert';

type ErrorBoundaryState = {
  hasError: boolean;
};

export class ErrorBoundary extends Component<{}, ErrorBoundaryState> {
  constructor(props: any) {
    super(props);
    this.state = { hasError: false };
  }
  static getDerivedStateFromError() {
    return { hasError: true };
  }
  render() {
    if (this.state.hasError) {
      return (
        <Alert
          alert={{
            message:
              "Sorry, something went wrong and we're experiencing some issues. Please try again.",
            type: AlertType.ERROR,
          }}
          showReload
        />
      );
    }

    return this.props.children;
  }
}
