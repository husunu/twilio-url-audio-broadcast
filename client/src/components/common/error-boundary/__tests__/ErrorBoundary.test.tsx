import React from 'react';
import { render, waitFor } from '@testing-library/react';

import { ErrorBoundary } from '../ErrorBoundary';

const ErrorComponent: React.FC<{ throwsError?: boolean }> = ({
  throwsError = false,
}) => {
  if (throwsError) {
    throw new Error('ErrorComponent');
  }
  return <div>No error</div>;
};

describe('ErrorBoundary', () => {
  test('shows children', async () => {
    const { container } = render(
      <ErrorBoundary>
        <ErrorComponent />
      </ErrorBoundary>
    );
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
  test('shows message when it catches an error', async () => {
    const spy = jest.spyOn(console, 'error');
    spy.mockImplementation(() => {});

    const { container } = render(
      <ErrorBoundary>
        <ErrorComponent throwsError />
      </ErrorBoundary>
    );
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });

    spy.mockRestore();
  });
});
