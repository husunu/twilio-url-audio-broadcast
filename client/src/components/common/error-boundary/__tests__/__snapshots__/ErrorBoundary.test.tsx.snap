// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary shows children 1`] = `
<div>
  <div>
    No error
  </div>
</div>
`;

exports[`ErrorBoundary shows message when it catches an error 1`] = `
<div>
  <div
    class="MuiPaper-root MuiAlert-root MuiAlert-standardError MuiPaper-elevation0"
    role="alert"
    style="align-items: center;"
  >
    <div
      class="MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message"
    >
      Sorry, something went wrong and we're experiencing some issues. Please try again.
    </div>
    <div
      class="MuiAlert-action"
    >
      <div
        class="MuiBox-root MuiBox-root-55"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Reload
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
